dali/pipeline/operators/util/external_source.h

```diff
@@ -92,21 +187,36 @@ class ExternalSource : public Operator<Backend> {
     return false;
   }
 
-  void RunImpl(Workspace<Backend> *ws) override;
+  void RunImpl(workspace_t<Backend> *ws) override;
 
-  string output_name_;
-  TensorList<Backend> tl_data_;
-  std::vector<Tensor<Backend>> t_data_;
-  bool data_in_tl_;
+  template<typename DataType>
+  void RecycleBuffer(DataType data,
+                     std::unique_ptr<detail::CudaEventWrapper> cuda_event = nullptr) {
+    if (cuda_event) {
+      cudaEventSynchronize(cuda_event->event);
+    }
+    std::lock_guard<std::mutex> busy_lock(busy_m_);
+    detail::RecycleHelper(std::move(data), tl_data_, t_data_);
+    if (cuda_event) {
+      cuda_events_.Recycle(std::move(cuda_event));
+    }
+  }
 
```

Comment:
I find the RecycleHelper more complicated than it should be

Why not:
```
private:
  void RecycleHelper(std::unique_ptr<std::vector<Tensor<CPUBackend>>>> data) {
    t_data_.Recycle(std::move(data));
  }

  void RecycleHelper(std::unique_ptr<TensorList<CPUBackend>> data) {
    tl_data_.Recycle(std::move(data));
  }
```
and then in RecycleBuffer you just call
```
RecycleHelper(std::move(data));
```
?

Reply:
Done